File: backend/app/services/ferramentas/processo.py
from sqlalchemy.orm import Session
from app.models import Processo, Movimento, ProcessoParte


SCHEMA_BUSCAR_PROCESSO = {
    "name": "buscar_processo",
    "description": "Busca dados completos de um processo judicial pelo numero CNJ. Retorna tribunal, classe, orgao julgador, partes e status.",
    "input_schema": {
        "type": "object",
        "properties": {
            "cnj": {
                "type": "string",
                "description": "Numero CNJ no formato NNNNNNN-DD.AAAA.J.TT.OOOO",
            },
        },
        "required": ["cnj"],
    },
}


def executar_buscar_processo(input_data: dict, db: Session) -> str:
    cnj = input_data.get("cnj", "").strip()
    if not cnj:
        return "Numero CNJ nao informado."
    processo = db.query(Processo).filter(Processo.cnj == cnj).first()
    if not processo:
        return f"Processo com CNJ {cnj} nao encontrado no sistema."

    partes = (
        db.query(ProcessoParte)
        .filter(ProcessoParte.processo_id == processo.id)
        .all()
    )

    linhas = [
        f"PROCESSO: {processo.cnj}",
        f"Tribunal: {processo.tribunal} ({processo.alias_tribunal})",
        f"Classe: {processo.classe_nome or 'N/A'}",
        f"Orgao Julgador: {processo.orgao_julgador or 'N/A'}",
        f"Grau: {processo.grau or 'N/A'}",
        f"Status: {processo.status}",
    ]
    if processo.data_ajuizamento:
        linhas.append(f"Ajuizamento: {processo.data_ajuizamento.strftime('%d/%m/%Y')}")

    if partes:
        linhas.append("")
        linhas.append("PARTES:")
        for p in partes:
            if p.cliente:
                linhas.append(f"  {p.papel.upper()}: {p.cliente.nome} (CPF/CNPJ: {p.cliente.cpf_cnpj})")
            else:
                linhas.append(f"  {p.papel.upper()}: (cliente removido, ID {p.cliente_id})")

    return "\n".join(linhas)


SCHEMA_LISTAR_MOVIMENTOS = {
    "name": "listar_movimentos",
    "description": "Lista os ultimos movimentos/andamentos de um processo. Retorna data, nome e resumo IA de cada movimento.",
    "input_schema": {
        "type": "object",
        "properties": {
            "processo_id": {
                "type": "integer",
                "description": "ID do processo no sistema",
            },
            "limite": {
                "type": "integer",
                "description": "Quantidade maxima de movimentos (padrao 20)",
            },
        },
        "required": ["processo_id"],
    },
}


def executar_listar_movimentos(input_data: dict, db: Session) -> str:
    processo_id = input_data.get("processo_id")
    if processo_id is None:
        return "Campo obrigatorio 'processo_id' nao informado."
    try:
        processo_id = int(processo_id)
    except (TypeError, ValueError):
        return "Campo 'processo_id' deve ser um numero inteiro."
    limite = min(max(int(input_data.get("limite", 20)), 1), 100)

    movimentos = (
        db.query(Movimento)
        .filter(Movimento.processo_id == processo_id)
        .order_by(Movimento.data_hora.desc())
        .limit(limite)
        .all()
    )

    if not movimentos:
        return f"Nenhum movimento encontrado para o processo ID {processo_id}."

    linhas = [f"MOVIMENTOS DO PROCESSO (ultimos {len(movimentos)}):"]
    for m in movimentos:
        data_str = m.data_hora.strftime("%d/%m/%Y %H:%M")
        linhas.append(f"  [{data_str}] {m.nome}")
        if m.resumo_ia:
            linhas.append(f"    Resumo: {m.resumo_ia}")

    return "\n".join(linhas)

File: frontend/lib/models/processo.dart
class Processo {
  final int id;
  final String cnj;
  final String numeroLimpo;
  final String tribunal;
  final String aliasTribunal;
  final int? classeCodigo;
  final String? classeNome;
  final String? orgaoJulgador;
  final String? grau;
  final DateTime? dataAjuizamento;
  final String status;
  final DateTime? ultimaVerificacao;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<ProcessoParte> partes;
  final List<Movimento> movimentos;

  Processo({
    required this.id,
    required this.cnj,
    required this.numeroLimpo,
    required this.tribunal,
    required this.aliasTribunal,
    this.classeCodigo,
    this.classeNome,
    this.orgaoJulgador,
    this.grau,
    this.dataAjuizamento,
    required this.status,
    this.ultimaVerificacao,
    required this.createdAt,
    required this.updatedAt,
    this.partes = const [],
    this.movimentos = const [],
  });

  factory Processo.fromJson(Map<String, dynamic> json) {
    return Processo(
      id: json['id'],
      cnj: json['cnj'],
      numeroLimpo: json['numero_limpo'],
      tribunal: json['tribunal'],
      aliasTribunal: json['alias_tribunal'],
      classeCodigo: json['classe_codigo'],
      classeNome: json['classe_nome'],
      orgaoJulgador: json['orgao_julgador'],
      grau: json['grau'],
      dataAjuizamento: json['data_ajuizamento'] != null
          ? DateTime.parse(json['data_ajuizamento'])
          : null,
      status: json['status'],
      ultimaVerificacao: json['ultima_verificacao'] != null
          ? DateTime.parse(json['ultima_verificacao'])
          : null,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      partes: (json['partes'] as List<dynamic>?)
              ?.map((e) => ProcessoParte.fromJson(e))
              .toList() ??
          [],
      movimentos: (json['movimentos'] as List<dynamic>?)
              ?.map((e) => Movimento.fromJson(e))
              .toList() ??
          [],
    );
  }
}

class ProcessoParte {
  final int id;
  final int processoId;
  final int clienteId;
  final String papel;

  ProcessoParte({
    required this.id,
    required this.processoId,
    required this.clienteId,
    required this.papel,
  });

  factory ProcessoParte.fromJson(Map<String, dynamic> json) {
    return ProcessoParte(
      id: json['id'],
      processoId: json['processo_id'],
      clienteId: json['cliente_id'],
      papel: json['papel'],
    );
  }
}

class Movimento {
  final int id;
  final int processoId;
  final int codigo;
  final String nome;
  final DateTime dataHora;
  final String? complementos;
  final String? resumoIa;
  final bool notificado;
  final DateTime createdAt;

  Movimento({
    required this.id,
    required this.processoId,
    required this.codigo,
    required this.nome,
    required this.dataHora,
    this.complementos,
    this.resumoIa,
    required this.notificado,
    required this.createdAt,
  });

  factory Movimento.fromJson(Map<String, dynamic> json) {
    return Movimento(
      id: json['id'],
      processoId: json['processo_id'],
      codigo: json['codigo'],
      nome: json['nome'],
      dataHora: DateTime.parse(json['data_hora']),
      complementos: json['complementos'],
      resumoIa: json['resumo_ia'],
      notificado: json['notificado'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }
}
